### DISCUSSION TITLE
Image hashes from API (issueidentifer)

### DISCUSSION THREAD

**Discussion by mizaki** [comictagger/comictagger#483] [Ideas]
Metron has now added pHash for its covers. My current thinking is:
- Add `hash` to ComicIssue (knowing the hash type will be required so maybe a JSON field? `{"type": "phash", "hash": "hexhash"}` Possibly support multiple hash types? Another option could be a string (array): `p_hexhash` or `a_hexhash` etc.)
- Get the hash for the local image
- If ComicIssue has a hash, ham it. If not do the usual download of image and hash

I'm wondering if the hash from the API should be ignored? If the list of possible issues does not match, maybe download the image and hash it ourselves to double check?

I'm also wondering if saving the hash of the local (to issue) or remote (to image cache) would be of any use?

When I comes to the hashing, should it be switched to the [imagehash](https://github.com/JohannesBuchner/imagehash) or implement the same in CTs? I notice that it requires a few deps that for CTs use case aren't really required?

**Comment by lordwelch** [comment]
The biggest problem is that the the hash algorithm is incredibly easy to change, if you resize before you turn to grayscale is different than turning grayscale before you resize. For example webp files there is a simple option that changes how it converts yuv/[cbcr] to RGB which then screws up the exact values when it resizes and converts to grayscale.

Even if ImageHash doesn't change, to keep the hash the same it depends on the version of PIL for resize and grayscale conversion and the version of libwebp or libjpeg or whichever library loads the image and if the library has a direct conversion to RGB or if PIL has to do it itself (PIL always converts images into RGB(A) images)

[cbcr]: https://en.wikipedia.org/wiki/YCbCr

**Reply by mizaki** [reply]
Sounds like you've done some testing? How much of a difference is there in those different scenarios? I thought one purpose of the process was to make things "close enough" for matching?

Is it significant enough to effect the hamming distance enough? 

I would presume that even with the variables, the distance is not enough to make a difference? We don't need the exact same number, just one close enough for confidence?

**Reply by lordwelch** [reply]
I don't have any conclusive testing but the Ahash which is the most prone to false-positives. In otherwords more images will have the same Ahash than Dhash or Phash. On my test of a single webp image with differences in resize/grayscale order a slightly different resize algorithm and incorrect color conversion it could not produce the same Ahash and the Dhash and Phash showed even more differences. After correcting the resize/grayscale order the resize algorithm and getting a little bit more accurate color conversion the Ahash matches and the Dhash hamming distance was ~5. Would they have matched? Yes. But this is the same exact file sent through both hash algorithm. When using a different file with actual differences it may be even less accurate. I plan on doing some better testing in the future but I want at least get webp to work exactly for my setup before I do that.

**Reply by lordwelch** [reply]
In this case I think we would just use the hash in place of the cover image, however CT currently does not have a phash implementation and I'm not sure I want to pull in numpy and scipy as further binary dependencies right now. I recently noticed that after some of my changes since I've taken over as maintainer for CT have drastically increased initial load time for at least the CLI and I want to see what I can do to fix that

**Reply by mizaki** [reply]
I did find this pure python for DCT https://gist.github.com/vilcans/8531160 and from what I remember from comparing the aHash the numpy isn't needed.

**Reply by lordwelch** [reply]
A hash and dhash are incredibly simple and don't need numpy in the slightest though it probably makes them a tad faster. If you are able to test it that would be great. There is some code for a perceptual algorithm that is commented out in imagehash.py I'm unsure if it is similar to the ImageHash library

**Comment by mizaki** [comment]
1. Now as `phash` returns the same result as IH lib (but in `int` form), it would make sense to have `ahash` do the same. IIRC the binary string was the same and it was only the last step that turned it into an `int` that changed things. Anything against a PR to change the `ahash` to return an `int` that can be `hex`'ed to be the same as IH `ahash`?
2. What format should the hash field on ComicIssue take? As above in the first message, JSON, list, dict? A dict such as {"ahash": int, "phash": int} might be best?

**Comment by killo3967** [comment]
I know I'm quite late to this conversation, but I want to share my experience. I've been mulling over this idea for a while, trying to find duplicate comics. 
The process I followed was to take the image, convert it to black and white, reduce it to 64x64 and calculate the phasing. 
I made a simple cmd script using ImageMagick. 
Then I calculated the Levenshtein separation and out of 150,000 there were 13 errors.
Then I tried ahash and dhash and the number of errors was very high.
As far as I can tell you, I hope it helps.

**Reply by lordwelch** [reply]
I have some old code to detect duplicates posted here https://github.com/comictagger/comictagger/discussions/575#discussioncomment-7845542

With the duplicates I found it was more effective to use the tagged data to actually find the duplicates and the hashes helped determine if they had similar quality images. I think I ended up finding a few comics that had tagged successfully but were incorrect
